adafruit_tlc59711/adafruit_tlc59711_multi.py

```diff
@@ -0,0 +1,1120 @@
+            raise ValueError(
+                "IoutR {} not in range: 2mA..{}mA"
+                "".format(IoutR, Ioclmax)
+            )
+        if not 0.0 <= IoutG <= Ioclmax:
+            raise ValueError(
+                "IoutG {} not in range: 2mA..{}mA"
+                "".format(IoutG, Ioclmax)
+            )
+        if not 0.0 <= IoutB <= Ioclmax:
+            raise ValueError(
+                "IoutB {} not in range: 2mA..{}mA"
+                "".format(IoutB, Ioclmax)
+            )
+        bcr = int((IoutR / Ioclmax) * 127)
+        bcg = int((IoutG / Ioclmax) * 127)
+        bcb = int((IoutB / Ioclmax) * 127)
+        if not 0 <= bcr <= 127:
+            raise ValueError(
+                "bcr {} not in range: 0..127"
+                "".format(bcr)
+            )
+        if not 0 <= bcg <= 127:
+            raise ValueError(
+                "bcg {} not in range: 0..127"
+                "".format(bcg)
+            )
+        if not 0 <= bcb <= 127:
+            raise ValueError(
+                "bcb {} not in range: 0..127"
+                "".format(bcb)
+            )
+        return (bcr, bcg, bcb)
+
+    ##########################################
+
+    def _debug_print_buffer(self):
+        indent = "  "
+        if self.chip_count == 1:
+            print("buffer: [", end="")
+            indent = ""
+        else:
+            print("buffer: [")
+        for index in range(self.chip_count):
+            print("{}{}: ".format(indent, index), end="")
+            # print()
+            # self._debug_print_tlc59711_bin()
+            # print()
+            # self._debug_print_tlc59711_hex()
+            self._debug_print_tlc59711()
+            if self.chip_count > 1:
+                print()
+        print("]")
+
+    def _debug_print_tlc59711_bin(self):
+        for index in range(len(self._buffer)):
+            print("{:08b}".format(self._buffer[index]), end="")
+
+    def _debug_print_tlc59711_hex(self):
+        for index in range(len(self._buffer)):
+            print("x{:02X}, ".format(self._buffer[index]), end="")
+
+    def _debug_print_tlc59711(self):
+        self._debug_print_tlc59711_header()
+        self._debug_print_tlc59711_ch()
+
+    def _debug_print_tlc59711_header(self):
+        # print(
+        #     "self.CHIP_BUFFER_HEADER_BYTE_COUNT",
+        #     self.CHIP_BUFFER_HEADER_BYTE_COUNT)
+        print("header: '", end="")
+        for index in range(self.CHIP_BUFFER_HEADER_BYTE_COUNT):
+            print("{:08b} ".format(self._buffer[index]), end="")
+        print("' ", end="")
+
+    def _debug_print_tlc59711_ch(self):
+        print("ch: [", end="")
+        for index in range(
+                self.CHIP_BUFFER_HEADER_BYTE_COUNT, len(self._buffer)
+        ):
+            print("x{:02X}, ".format(self._buffer[index]), end="")
+        print("]", end="")
+
+    ##########################################
+
+    def _get_32bit_value_from_buffer(self, buffer_start):
```

Comment:
Use struct instead.

Reply:
And same for routines below.

Reply:
The struct docs for CPython are here: https://docs.python.org/3/library/struct.html We implement most of the same API. You'll want pack_into.

Reply:
@tannewt thanks! i forgot about this... 
i implemented it and found that with the struct call in some cases it is very slow.
so after i *cleaned up* all my calls i readded the discreet set in most again for speed reasons.
i have documented this in every function with example times.